jina/peapods/compoundpod/__init__.py

```diff
@@ -0,0 +1,384 @@
+__copyright__ = "Copyright (c) 2020 Jina AI Limited. All rights reserved."
+__license__ = "Apache-2.0"
+
+import argparse
+from argparse import Namespace
+from typing import Optional, Dict, List, Union, Set
+
+from .helper import (
+    _set_pod_args,
+    _set_after_to_pass,
+    _copy_to_head_args,
+    _copy_to_tail_args,
+    _fill_in_host,
+)
+from .. import Pea
+from .. import Pod
+from .. import BasePod
+from ...enums import *
+
+
+class CompoundPod(BasePod):
+    """A BasePod is a immutable set of peas, which run in parallel. They share the same input and output socket.
+    Internally, the peas can run with the process/thread backend. They can be also run in their own containers
+    """
+
+    def __init__(self, args: Union['argparse.Namespace', Dict], needs: Set[str] = None):
+        """
+        :param args: arguments parsed from the CLI
+        :param needs: pod names of preceding pods, the output of these pods are going into the input of this pod
+        """
+        super().__init__()
+
+        self.args = args
+        self._set_conditional_args(self.args)
+        self.needs = (
+            needs if needs else set()
+        )  #: used in the :class:`jina.flow.Flow` to build the graph
+
+        self.replica_list = []  # type: List['Pod']
+        self.peas = []  # type: List['Pea']
+        self.is_head_router = False
+        self.is_tail_router = False
+        self.deducted_head = None
+        self.deducted_tail = None
+
+        if isinstance(args, Dict):
+            # This is used when a Pod is created in a remote context, where peas & their connections are already given.
+            self.replicas_args = args
+        else:
+            self.replicas_args = self._parse_args(args)
+
+    @property
+    def role(self) -> 'PodRoleType':
+        """
+        Return the role of this :class:`CompoundPod`.
+
+        :return: role type
+        """
+        return self.args.pod_role
+
+    @property
+    def is_singleton(self) -> bool:
```

Comment:
I think `is_singleton` may not make sense in `CompoundPod`?

Reply:
It would make sense if we allow that the number of replicas is == 1.
But since we don't support this case, I will remove it, thanks.